Retrieve unannotated Git tags too

brianhelba opened the pull request:
Unannotated Git tags should be considered alongside annotated tags when retrieving a version from Git.

Compared to annotated tags, unannotated tags do not contain a distinct commit message or metadata, but neither are needed to retrieve version name.

See: https://git-scm.com/docs/git-describe#Documentation/git-describe.txt---tags

brianhelba commented:
@isaacs This is ready for review. Please take a look.

isaacs commented on lib/retrieve-tag.js:
```diff
@@ -2,7 +2,7 @@ const {spawn} = require('@npmcli/git')
 
 module.exports = async opts => {
-  const tag = (await spawn(['describe', '--abbrev=0'], opts)).stdout.trim()
+  const tag = (await spawn(['describe', '', '--abbrev=0'], opts)).stdout.trim()
```
Did you mean for this to be an empty string?  The discussion makes it seem like this second arg should be `'--tags'`, no?

isaacs replied:
`git describe "" --abbrev=0` says:

```
$ git describe "" --abbrev=0
fatal: Not a valid object name
```

brianhelba replied:
Yikes, oops. I used GitHub's web IDE to make this change, and I did not notice that made the commit with the wrong content.

I just force-pushed the correct code.

brianhelba committed:
Retrieve unannotated Git tags too

Unannotated Git tags should be considered alongside annotated tags when
retrieving a version from Git.

Compared to annotated tags, unannotated tags do not contain a distinct
commit message or metadata, but neither are needed to retrieve version name.

See: https://git-scm.com/docs/git-describe#Documentation/git-describe.txt---tags

evocateur commented:
I don't think `libnpmversion` should be allowing lightweight tags to masquerade as release tags. Moreover, `libnpmversion` itself only creates "annotated" tags, which the [git tag documentation](https://git-scm.com/docs/git-tag) itself indicates are appropriate for a "release":

> Annotated tags are meant for release while lightweight tags are meant for private or temporary object labels. For this reason, some git commands for naming objects (like `git describe`) will ignore lightweight tags by default.

I can't think of a reason lightweight tags would be preferable for this sort of behavior.

ljharb commented:
I’ve never known there was a difference, and probably half my projects’ versions have unannotated release tags. I would be very surprised if the majority of package authors intentionally differentiated tags that way.

evocateur commented:
@ljharb `npm version` itself has never created lightweight tags, always annotating them (implicitly) with the `-m` flag.

> I would be very surprised if the majority of package authors intentionally differentiated tags that way.

I'd be very surprised if the majority of package authors _manually created their release tags_ instead of using `npm version`. If somebody does take it upon themselves to create release tags manually, they should be capable of understanding the difference. Since `git describe` is only used when the `'from-git'` argument is used, it should be straightforward enough to throw an error when no version was recovered and document appropriately.

As someone who has dealt with release tags and the mechanics of npm versioning for quite some time, I would like to preserve the (in my opinion, _correct_) behavior of `npm version` (and by extension, `libnpmversion`) that ensures it creates properly annotated tag objects (unless completely avoided by `--no-git-tag-version` / `{ gitTagVersion: false }`, ofc). Opening it up to lightweight tags opens up a whole new universe of random temporary tags (or tags that have nothing to do with releases, since there is no filtering by pattern whatsoever, just the "most recent"), which is an entire bug farm unto itself.

ljharb commented:
`npm version` doesn't work well with including a changelog in the tagged release commit, and a great many packages that aren't mine manually create git tags - you can tell by the `v` in the tag message content.

I didn't understand the difference when manually tagging a very high number of package-versions, for the better part of a decade. I don't think it's reasonable to expect authors to understand every bit of git minutiae. I'm totally fine with npm only *creating* tagged versions, but i think it should retrieve every semver-ish tag, annotation or not.

Are you concerned with cases where a lightweight tag is named with a valid semver range?

evocateur commented:
> a great many packages that aren't mine manually create git tags - you can tell by the v in the tag message content.

Do those maintainers use `npm version from-git`? Are they even _aware_ of it? If they are, why are they bothering when they've already done half the work of the command already?

(You can customize the commit message by passing [`-m`/`--message`](https://docs.npmjs.com/cli/v6/commands/npm-version#description), so that's not a very strong indication that the tags are manually created)

> Are you concerned with cases where a lightweight tag is named with a valid semver range?

I'm concerned that the current `from-git` call, if modified as this PR intends, will pick up unrelated lightweight tags (`"testing"`, `"in-prod-now-lol"`, etc) and fail because they do not match the pattern. That's a dramatic, easily-avoided footgun.

ljharb commented:
What if in addition to annotated tags, it also picked up lightweight tags that were a valid semver version?

evocateur commented:
If `--tags` is passed to `git describe`, the only way to avoid lightweight false-positives is to pass [`--match='*.*.*'`](https://git-scm.com/docs/git-describe#Documentation/git-describe.txt---matchltpatterngt) as well, instead of relying on an extremely permissive post-hoc regex filter.

ljharb commented:
presumably it could be done in two calls, one that only includes annotated tags, and another that filters out all tags afterwards with `semver`, and then those two be combined in a Set later?

evocateur commented:
A single call would be sufficient, as long as all the arguments are passed.

```sh
git describe --abbrev=0 --tags --match='v*.*.*'
```

The idea is that if we're going to accept lightweight tags, ensure they match the configured pattern and the first one to be resolved by `git describe` will be the most recent, regardless of annotation.

(While we're at it, we should also respect `opts.tagVersionPrefix` when configuring the `--match` value, instead of hard-coding `v` as in the example above)

ljharb commented:
That pattern wouldn’t include prerelease tags, and would include invalid semver versions.

evocateur commented:
<img width="499" alt="git describe matching prerelease tags" src="https://user-images.githubusercontent.com/5605/107430452-0852f000-6ada-11eb-8a3b-88c7156c05d7.png">

That's describing a tag created by `npm version prerelease`, btw. What kind of "prerelease tags" are you thinking of?

As for "invalid semver versions", I think we're just going to have to trust people to not write tags like `vYEET.YEET.YEET`?

ljharb commented:
ah ok, seems like it will match them, that's good. I don't see why there can't be a filter using `semver.isValidRange` tho, to keep the yeets away :-p

evocateur commented:
Oh, for sure, we'd want to ensure the matched tag was valid, once retrieved from `git describe`.

ljharb commented:
@wraithgar was this intended to be closed?

wraithgar commented:
No, but there was no way to avoid this because of the change of the default branch.  OP will have to re-apply this PR to `main` now, we can't reopen it from this end.

ljharb commented:
I'm confused; github automatically retargets these PRs through the "change default branch" interface. Why did this one not work properly?

wraithgar commented:
I haven't the foggiest. I can't retarget this PR, it's not even an option now.

wraithgar commented:
https://github.com/npm/libnpmversion/pull/12
